src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -50,6 +51,27 @@ public static ImmutableArray<TextDocument> GetTextDocuments(this Solution soluti
             return documentIds.SelectAsArray(id => getDocument(solution, id));
         }
 
+        public static (Document?, Solution) GetDocumentAndSolution(this ILspSolutionProvider provider, TextDocumentIdentifier? textDocument, string? clientName)
+        {
+            var solution = provider.GetCurrentSolutionForMainWorkspace();
+            if (textDocument != null)
+            {
+                var document = provider.GetDocument(textDocument, clientName);
+                var solutionOfDocument = document?.Project.Solution;
+
+                return (document, solutionOfDocument ?? solution);
+            }
+
+            return (null, solution);
+        }
+
+        public static RequestContext CreateRequestContext(this ILspSolutionProvider provider, TextDocumentIdentifier textDocument, ClientCapabilities clientCapabilities, string? clientName = null)
```

Comment:
If this is only used by the TS handlers, maybe we should just make this a static method there so we can remove it when we remove the rest of the TS shims.

Reply:
Do you mean create a new static class in TypeScriptHandlerShims.cs for this, just so deleting that file deletes everything? Can certainly do that.

Reply:
yup, just so we don't forget to delete it